### PR TITLE
Add batch_ndims argument in Predictive and log_likelihood

### PR DESCRIPTION
**Pull request by fehiepsi**
Addresses #645 and #639.

This PR also adds a utility `soft_vmap` to avoid duplicated code in Predictive and log_likelihood. It might also be useful for other purposes. @martinjankowiak you might find `soft_vmap`, which is adapted from your code, useful for your experiment. :D

### CODE REVIEW

File: numpyro/util.py [hunk 1]
```diff
@@ -298,3 +298,29 @@ def unravel_pytree(arr):
         return tree_unflatten(treedef, unravel_list(arr))
 
     return flat, unravel_pytree
+
+
+def soft_vmap(fn, xs, batch_shape, chunk_size=None):
```
Comment (martinjankowiak) [inline numpyro/util.py:303]: @fehiepsi could you please add a docstring for this? i'd like to better understand what it's doing...

Comment (fehiepsi) [reply numpyro/util.py:303]: Sure, I'll add. This basically computes `fn(xs)` by vectorizing `fn(x)` over `batch_shape` (i.e. `xs.shape = batch_shape + 'x.shape'`). The optional `chunk_size` allows the job taken in chunk: by spliting `xs` into chunks `xs_1`,..., `xs_n`, computing `fn(xs_1)`, ... `fn(xs_n)` using `vmap`, and finally merging the results. `chunk_size=None` means that there is only 1 chunk, which is the same as `vmap(fn)(x)`. `chunk_size=1` means that we don't use `vmap` at all, which is the same as `map(fn)(x)`.

The difference w.r.t. your script is `batch_shape` argument, which has length 1 in your script but has arbitrary length here (to allow calculating `num_chains x num_samples x ...`, which has length 2, forms).

File: numpyro/util.py [hunk 2]
```diff
@@ -298,3 +298,42 @@ def unravel_pytree(arr):
         return tree_unflatten(treedef, unravel_list(arr))
 
     return flat, unravel_pytree
+
+
+def soft_vmap(fn, xs, batch_ndims=1, chunk_size=None):
+    """
+    Vectorizing map a function `fn` over `batch_ndims` leading axes of `xs`.
+    The job is splitted into chunks to reduce memory usages.
```
Comment (neerajprad) [inline numpyro/util.py:306]: nits: 
Vectorizing map that maps a function `fn` ..
splitted --> split

Maybe we can replace the last line with: This uses `jax.vmap` over smaller chunks of the batch dimension to keep memory usage constant.

File: test/test_infer_util.py [hunk 3]
```diff
@@ -130,14 +131,21 @@ def test_prior_predictive():
     assert predictive_samples["obs"].shape == (100,) + data.shape
 
 
-def test_log_likelihood():
+@pytest.mark.parametrize('batch_shape', [(), (100,), (2, 50)])
+def test_log_likelihood(batch_shape):
     model, data, _ = beta_bernoulli()
-    samples = Predictive(model, return_sites=["beta"], num_samples=100)(random.PRNGKey(1))
-    loglik = log_likelihood(model, samples, data)
+    samples = Predictive(model, return_sites=["beta"], num_samples=200)(random.PRNGKey(1))
```
Comment (neerajprad) [inline test/test_infer_util.py:137]: What will happen if we provide `batch_ndims=2` here?

File: numpyro/infer/util.py [hunk 4]
```diff
@@ -488,21 +514,48 @@ def get_samples(self, rng_key, *args, **kwargs):
...
-        model_trace = trace(substitute(model, samples)).get_trace(*args, **kwargs)
+        substituted_model = substitute(model, samples) if isinstance(samples, dict) else model
+        model_trace = trace(substituted_model).get_trace(*args, **kwargs)
         return {name: site['fn'].log_prob(site['value']) for name, site in model_trace.items()
                 if site['type'] == 'sample' and site['is_observed']}
 
-    return vmap(single_loglik)(posterior_samples)
+    prototype_site = batch_shape = None
+    for name, sample in posterior_samples.items():
+        if batch_shape is not None and sample.shape[:batch_ndims] != batch_shape:
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/infer/util.py:547]: Sorry, outdated comment. This is being inferred by the first element in the dict, so should be fine.

File: numpyro/util.py (hunk 1 above)
Comment (fehiepsi) [reply numpyro/util.py]: I just add the docstring.

File: test/test_infer_util.py (hunk 3 above)
Comment (fehiepsi) [reply test/test_infer_util.py:137]: I believe the output will have shape `(1, 200, ...)` but let me test it.

Comment (fehiepsi) [reply test/test_infer_util.py:140]: That's true, I just add a test for that in `test_prior_predictive`.

File: numpyro/distributions/transforms.py [hunk 5]
```diff
@@ -371,8 +371,8 @@ def log_abs_det_jacobian(self, x, y, intermediates=None):
 
 class MultivariateAffineTransform(LowerCholeskyAffine):
     def __init__(self, loc, scale_tril):
-        warnings.warn(DeprecationWarning,
-                      "MultivariateAffineTransform is renamed to LowerCholeskyAffine.")
+        warnings.warn("MultivariateAffineTransform is renamed to LowerCholeskyAffine.",
+                      DeprecationWarning)
```
Comment (neerajprad) [inline numpyro/distributions/transforms.py:375]: Sorry, I might have missed this earlier.`DeprecationgWarning` will be swallowed by application code. We should use `FutureWarning` instead. See https://docs.python.org/3/library/warnings.html#warning-categories.

File: numpyro/infer/initialization.py [hunk 6]
```diff
@@ -32,8 +32,8 @@ def init_to_median(site=None, num_samples=15):
 
 
 def init_to_prior(site=None):
-    warnings.warn(DeprecationWarning,
-                  "`init_to_prior` strategy is renamed to `init_to_sample`.")
+    warnings.warn("`init_to_prior` strategy is renamed to `init_to_sample`.",
```
Comment (neerajprad) [inline numpyro/infer/initialization.py:35]: same here about DeprecationWarning.

File: numpyro/util.py (hunk 2 above)
Comment (fehiepsi) [reply numpyro/util.py]: Thanks for suggesting!

File: numpyro/distributions/transforms.py (hunk 5 above)
Comment (fehiepsi) [reply numpyro/distributions/transforms.py]: Thanks! This explains why I didn't see the warnings when I used `predictive.get_samples(...)`. :D

**Pull request merged**